sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -246,304 +247,620 @@ abstract class RapidsShuffleThreadedWriterBase[K, V](
+    val partialFileInputs = partialFiles.map { pf =>
+      new FileInputStream(pf.file)
+    }
+
+    try {
+      // For each partition, copy data from all partial files in order
+      (0 until numPartitions).foreach { partitionId =>
+        val partWriter = finalWriter.getPartitionWriter(partitionId)
+        val maybeOutputChannel: Optional[WritableByteChannelWrapper] =
+          partWriter.openChannelWrapper()
+
+        if (maybeOutputChannel.isPresent && transferToEnabled) {
+          // Use channel-based fast copy with NIO
+          val outputChannel = maybeOutputChannel.get()
+          withResource(outputChannel) { _ =>
+            partialFiles.zip(partialFileInputs).foreach {
+              case (partialFile, inputStream) =>
+                val partitionLength = partialFile.partitionLengths(partitionId)
+                if (partitionLength > 0) {
+                  val inputChannel = inputStream.getChannel
+                  val currentPos = inputChannel.position()
+                  Utils.copyFileStreamNIO(
+                    inputChannel,
+                    outputChannel.channel,
+                    currentPos,
+                    partitionLength)
+                  // Ensure position advances (copyFileStreamNIO should do this)
+                  inputChannel.position(currentPos + partitionLength)
+                }
             }
-          } else {
-            writePartitionedDataWithStream(file, partWriter)
           }
-          file.delete()
+        } else {
+          // Use stream-based copy (no channels, pure streams)
+          withResource(partWriter.openStream()) { os =>
+            partialFiles.zip(partialFileInputs).foreach {
+              case (partialFile, inputStream) =>
+                val partitionLength = partialFile.partitionLengths(partitionId)
+                if (partitionLength > 0) {
+                  // Read sequentially from current stream position
+                  // No skip/offset needed - stream position auto-advances!
+                  val buffer = new Array[Byte](fileBufferSize)
+                  var remainingBytes = partitionLength
+                  while (remainingBytes > 0) {
+                    val bytesToRead = math.min(remainingBytes, buffer.length).toInt
+                    val bytesRead = inputStream.read(buffer, 0, bytesToRead)
+                    if (bytesRead > 0) {
+                      os.write(buffer, 0, bytesRead)
+                      remainingBytes -= bytesRead
+                    } else {
+                      throw new IOException(
+                        s"Unexpected end of file reading partition $partitionId")
+                    }
+                  }
+                }
+            }
+          }
+        }
+      }
+    } finally {
+      // Close all partial file streams
+      partialFileInputs.foreach { in =>
+        try {
+          in.close()
+        } catch {
+          case _: Exception => /* ignore */
```

Comment:
instead of just ignoring them can we at least log these? I don't expect the exception case to be common, but if it does show up it might be nice to have the information so we can debug it.